include/IECoreScene/MeshAlgo.h

```diff
@@ -47,8 +47,24 @@ namespace IECoreScene
 namespace MeshAlgo
 {
 
-/// Calculate the surface tangent vectors of a mesh primitive.
+enum MeshTangentsMode
+{
+	UV,
+	FirstEdge,
+	TwoEdges,
+	PrimitiveCentroid
+};
+
+/// TODO: remove this compatibility function:
 IECORESCENE_API std::pair<PrimitiveVariable, PrimitiveVariable> calculateTangents( const MeshPrimitive *mesh, const std::string &uvSet = "uv", bool orthoTangents = true, const std::string &position = "P" );
+/// Calculate the surface tangent vectors of a mesh primitive based on UV information
+IECORESCENE_API std::pair<PrimitiveVariable, PrimitiveVariable> calculateTangentsFromUV( const MeshPrimitive *mesh, const std::string &uvSet = "uv", const std::string &position = "P", bool orthoTangents = true, bool leftHanded = false );
+/// Calculate the surface tangent vectors of a mesh primitive based on the first neighbor edge
+IECORESCENE_API std::pair<PrimitiveVariable, PrimitiveVariable> calculateTangentsFromFirstEdge( const MeshPrimitive *mesh, const std::string &position = "P", const std::string &normal = "N", bool orthoTangents = true, bool leftHanded = false );
```

Comment:
Sorry, this comment somehow got dropped from my review...

@johnhaddon what do you think about the naming here? I found it a bit redundant at first glance, as the 3 new methods could be simplified to

`calculateTangents( const MeshPrimitive *mesh, MeshTangentsMode mode, const std::string &position = "P", const std::string &normal = "N", bool orthoTangents = true, bool leftHanded = false )`

but it'd be a bit odd to just have `calculateTangentsFromUV` on its own... is there some c++11 variable arg magic that can solve this for us? Or do you prefer the more verbose naming?

Reply:
The function signatures are already pretty complex, so trying to squeeze them all into one (with a uv argument that is redundant for most modes and a normal argument that is ignored by uv mode) would be confusing to my mind. But having a MeshTangentsMode that isn't used anywhere in Cortex doesn't make any sense either, so we should either remove it, remove the UV value from it, or have a single function that can do all modes. I don't feel super strongly about it, but I'd be happy enough to keep the individual methods as we have now...